### PR TITLE
Support DirEntry metadata calls in miri

### PR DESCRIPTION
**Pull request by SUPERCILEX**
This should work as it uses lstat64 which is supported here: ~https://github.com/rust-lang/miri/blob/d9ad25ee4bbd9364c498959cdc82b5fa6c41e63c/src/shims/unix/macos/foreign_items.rs#L42~ just noticed that's macos, linux would be using statx: https://github.com/rust-lang/miri/blob/86f0e63b21721fe2c14608644f467b9cb21945eb/src/shims/unix/linux/foreign_items.rs#L112

The failing syscall is `dirfd`, so maybe that should actually be added to the shims?

### CODE REVIEW

**Comment by RalfJung** [comment]
I can't approve miri-specific library hacks, that needs someone from the libs team.
Cc @thomcc 

However, from the Miri side, this is lacking tests that ensure that this actually works.

> The failing syscall is dirfd, so maybe that should actually be added to the shims?

The problem is that dirfd is not exposed by the standard library, so Miri has no good way of implementing that shim in a cross-platform way.

**Comment by thomcc** [comment]
> The problem is that dirfd is not exposed by the standard library, so Miri has no good way of implementing that shim in a cross-platform way.

This would bother me more if we didn't already have both code-paths -- maintaining the `miri` in the cfg does not worry me.

It is unfortunate from the perspective of having slightly inaccurate emulation on these targets, though.

> However, from the Miri side, this is lacking tests that ensure that this actually works.

Hmm, this is a good point I guess. Would adding a test for this function in libstd be sufficient here?

**Comment by RalfJung** [comment]
Which libstd API function is enabled by this? The test should be added in `src/tools/miri/tests/pass-dep/shims/fs.rs`.

**Comment by thomcc** [comment]
Thanks. Looks good to me if @RalfJung doesn't want any changes to the tests.

**Comment by oli-obk** [comment]
@bors r=thomcc

Looks good to me

File: src/tools/miri/tests/pass-dep/shims/fs.rs [hunk 1]
```diff
@@ -404,6 +404,14 @@ fn test_directory() {
     let mut file_names = dir_iter.map(|e| e.unwrap().file_name()).collect::<Vec<_>>();
     file_names.sort_unstable();
     assert_eq!(file_names, vec!["test_file_1", "test_file_2"]);
+    // Test that read_dir metadata calls succeed
+    assert_eq!(
+        &[true, true],
```
Comment (RalfJung) [inline src/tools/miri/tests/pass-dep/shims/fs.rs:409]: Ideally we'd also make sure it returns `false` non directories.

Comment (SUPERCILEX) [reply src/tools/miri/tests/pass-dep/shims/fs.rs:409]: Sounds good, created https://github.com/rust-lang/rust/pull/103258

**Pull request merged**